bridges/snowbridge/pallets/system-frontend/src/lib.rs

```diff
@@ -230,14 +255,60 @@ pub mod pallet {
 				.saturating_add(T::BackendWeightInfo::transact_add_tip())
 		)]
 		pub fn add_tip(origin: OriginFor<T>, message_id: MessageId, asset: Asset) -> DispatchResult {
-			ensure_signed(origin)?;
-			// Check tip is DOT
-
-			// Convert to ether
+			let who = ensure_signed(origin)?;
+
+			let ether_location = T::EthereumLocation::get();
+			let (tip_asset_location, tip_amount) = match asset {
+				Asset {
+					id: AssetId(ref loc),
+					fun: Fungibility::Fungible(amount),
+				} => (loc, amount),
+				_ => return Err(Error::<T>::UnsupportedAsset.into()),
+			};
+
+			// Swap tip asset to ether
+			let swap_path = vec![(*tip_asset_location).clone(), ether_location.clone()];
+			let who_location = Self::account_to_location(who.clone())?;
+
+			let balance_before = T::ForeignToken::balance(ether_location.clone(), &who);
+			T::Swap::swap_exact_tokens_for_tokens(
+				who.clone(),
+				swap_path,
+				tip_amount,
+				None,
+				who.clone(),
+				true,
+			)?;
+			let balance_after = T::ForeignToken::balance(ether_location.clone(), &who);
+			let ether_gained = balance_after.saturating_sub(balance_before);
+
+			// Burn the ether
+			let ether_asset = Asset::from((ether_location.clone(), ether_gained));
+			let xcm_context = XcmContext {
+				origin: None,
+				message_id: Default::default(),
+				topic: None,
+			};
+			T::AssetTransactor::can_check_out(&who_location, &ether_asset, &xcm_context).map_err(|_| Error::<T>::WithdrawError)?;
+			T::AssetTransactor::check_out(&who_location, &ether_asset, &xcm_context);
+			T::AssetTransactor::withdraw_asset(&ether_asset, &who_location, None).map_err(|_| Error::<T>::WithdrawError)?;
+
+			// Send the tip details to BH to be allocated to the reward in the Inbound/Outbound pallet
+			let dest = T::BridgeHubLocation::get();
+			let call = Self::build_add_tip_call(message_id.clone(), ether_gained);
+			let remote_xcm = Self::build_remote_xcm(&call);
+			let local_pallet_origin: Location = T::PalletLocation::get().into();
 
-			// Burn DOT
+			let xcm_message_id =
+				Self::send_xcm(local_pallet_origin, dest.clone(), remote_xcm.clone())
+					.map_err(|error| Error::<T>::from(error))?;
 
-			// Call system v2::add-tip
+			Self::deposit_event(Event::<T>::RewardTipAdded {
```

Comment:
We should just emit a generic `MessageSent`, like `register_token`. Does `swap_exact_tokens_for_tokens` emit its own events?

Reply:
Cool, will do. Yes, it emits a detailed [SwapExecuted](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/asset-conversion/src/lib.rs#L766) event.

Reply:
What I meant is that you should just emit `Self::deposit_event(Event::<T>::MessageSent {}`) like we do in `register_token`.

Reply:
The MessageSent event will not contain any details about the MessageId provided. The `asset` field in the event also does not really match the data we use in `add_tip`. I could massage the tip into an Asset and use that in the event, though.

Reply:
Addressed in https://github.com/vgeddes/polkadot-sdk/pull/23/commits/cdc1cc166d0e9c5642575fb4c44717aecd27bf4a.